doc/client_channel_specification.md

```diff
@@ -0,0 +1,809 @@
+will report CONNECTING state with a picker that queues all picks.
+
+As soon as one of the subchannels becomes connected, pick\_first will
+report READY with a picker that always returns complete picks with that
+subchannel.  It will unref all of the other subchannels.  (But note that
+if pick\_first is reporting [health state](#subchannel-health-state)
+to a parent policy like round\_robin, it may instead report
+TRANSIENT\_FAILURE.)
+
+It will stay in READY state until either the connection fails or it
+receives a resolver update that does not include the address of the
+subchannel it is connected to.  In either case, it will unref the
+subchannel and report IDLE with a picker that calls the LB policy's
+`ExitIdle()` method when it receives the first pick and then queues
+all picks.
+
+In IDLE state, when `ExitIdle()` is called, pick\_first will create a new
+list of subchannels using the last address list it saw from the resolver
+and start connecting to them, exactly as it does when it is first created.
+It will therefore report CONNECTING state with a picker that queues picks.
+
+If pick\_first fails to connect to all subchannels, it will report
+TRANSIENT\_FAILURE state with a picker that fails picks.  The RPC
+status that is used to fail the picks should include the error from the
+most recent subchannel connection failure.  It will then keep trying to
+connect to all subchannels indefinitely, remaining in TRANSIENT\_FAILURE
+state until it succeeds in establishing a connection, at which point
+it transitions to READY state.  In particular, note that it will not
+transition to CONNECTING in between TRANSIENT\_FAILURE and READY, even
+though the underlying subchannels do so.
+
+The policy will request re-resolution via the helper when it initially
+reports TRANSIENT\_FAILURE, and then again every time it counts a number
+of connection attempt failures equal to the number of subchannels.
+It will also request re-resolution when a selected subchannel transitions
+from READY to IDLE.
+
+If pick\_first receives a resolver update with an empty address list,
+it will report TRANSIENT\_FAILURE and return an error telling its parent
+that it rejected the update.  The channel will pass this error back to
+the resolver with the health status callback.  For polling resolvers,
+this will cause the resolver to re-resolve with appropriate backoff.
+
+This is the default LB policy if no policy is set via the channel
+API or the [service config](#service-config).  It is also used
+as the universal leaf policy for dualstack support, as per [gRFC
+A61](https://github.com/grpc/proposal/blob/master/A61-IPv4-IPv6-dualstack-backends.md)
+-- i.e., policies like [round\_robin](#round_robin) that balance across
+endpoints will delegate to a pick\_first child policy to choose an
+address within each endpoint.
+
+#### round\_robin
+
+The round\_robin policy spreads load across a set of endpoints by
```

Comment:
We have these detailed explanations of PF & RR - should we have something for DNS?

Specifically, this paragraph & the next talk about endpoints.  It seems like somewhere we should document that DNS outputs one endpoint for every address.

Reply:
Good catch.  Added a note about this in the DNS resolver section above.